docs/sources.md

```diff
@@ -0,0 +1,268 @@
+
+In that directory, create any files you need. However, the following must be exported as publicly available:
+
+```go
+func NewSource(opt Opt) (source.Source, error)
+```
+
+The above must return a `source.Source` interface, whose definition is [here](../source/manager.go#L13):
+
+```go
+type Source interface {
+	ID() string
+	Resolve(ctx context.Context, id Identifier, sm *session.Manager, vtx solver.Vertex) (SourceInstance, error)
+}
+```
+
+The function `ID()` must return the type name created earlier in `types.go`.
+
+The function `Resolve()` must return a `SourceInstance`, which is the implementation used to perform functions on a specific source identifier.
+The `id` will give the unique identifier for the source to be retrieved. For example, if the source is `docker-image://docker.io/library/nginx:latest`,
+then it will use the source registered to handle `docker-image`, and when `Resolve()` is called, it will be passed `docker.io/library/nginx:latest`
+as the value of `id`.
+
+The returned value from `Resolve()` must implement [SourceInstance](../source/manager.go#L18):
+
+```go
+type SourceInstance interface {
+	CacheKey(ctx context.Context, g session.Group, index int) (string, string, solver.CacheOpts, bool, error)
+	Snapshot(ctx context.Context, g session.Group) (cache.ImmutableRef, error)
+}
+```
+
+A `SourceInstance` is returned by `Resolve()`, so it is assumed already to know what its identifier is, e.g. `docker.io/library/nginx:latest`.
+
+The two functions in the signature of the `SourceInstance` are expected to perform the following functions.
+
+#### `CacheKey`
+
+`CacheKey()` gives the solver a key that uniquely identifies the instance of the operation (including its arguments and inputs).
+So for a container image source, it would basically amount to the image digest (plus some other metadata).
+
+The purpose of `CacheKey()` is for the solver to be able to ask the source, "I have this `id`, what unique cache key is there for it?"
+If the response is identical to the one already in cache, the solver knows it does not need to retrieve the data again, and simply can
+use what is in its cache. If it is different or does not exist, then it knows it needs to ask the source for the content referred to by the
+`id`
+
+This is expected to be idempotent, a read-only activity that returns a unique cache key.
+
+The arguments to `CacheKey()` are:
+
+* `ctx context.Context`: A golang [context](https://pkg.go.dev/context#Context).
+* `g session.Group`: The [session.Group](https://pkg.go.dev/github.com/moby/buildkit@v0.10.1/session#Group) contains parameters to identify the specific buildkit client calling this build. These may be useful for things like credentials.
+* `index int`: The `CacheKey()` can be called multiple times for the same source item, each indicating that it is a separate subitem. The `index` indicates which iteration of call this is. Generally, if the `index > 0` then it is complete.
```

Comment:
This part really threw me for a loop. The `solver` doesn't know more than, "here is an ID, give me a cache key". How can it know of multiple index integers? When does it change them? What is its meaning?